storage/mysql/log_storage.go

```diff
@@ -274,12 +276,23 @@ func (t *logTX) QueueLeaves(leaves []trillian.LogLeaf) error {
 		}
 	}
 
+	// If the log does not allow duplicates we prevent the insert of such a leaf from
+	// succeeding. If duplicates are allowed multiple sequenced leaves will share the same
+	// leaf data in the database.
+	var insertSQL string
+
+	if t.ls.allowDuplicates {
+		insertSQL = insertUnsequencedLeafSQL
+	} else {
+		insertSQL = insertUnsequencedLeafSQLNoDuplicates
+	}
+
 	for _, leaf := range leaves {
 		// Create the unsequenced leaf data entry. We don't use INSERT IGNORE because this
 		// can suppress errors unrelated to key collisions. We don't use REPLACE because
 		// if there's ever a hash collision it will do the wrong thing and it also
 		// causes a DELETE / INSERT, which is undesirable.
-		_, err := t.tx.Exec(insertUnsequencedLeafSQL, t.ls.logID.TreeID,
+		_, err := t.tx.Exec(insertSQL, t.ls.logID.TreeID,
```

Comment:
If we're going to do it for potential efficiency gains we should probably look at the other transaction level prepared statements. I don't think that belongs in this PR.